functions-continued/variables-are-references.md

```diff
@@ -0,0 +1,648 @@
+print("value of berries:", berries)
+print("value of melons:", melons)
+```
+
+When we run the above code, we may see output like this:
+
+```
+Before modifying berries
+id of berries: 4369938880
+id of melons: 4369938880
+value of melons: ['hearts', 'stars', 'horseshoes']
+After modifying berries
+id of berries: 4369938880
+id of melons: 4369938880
+value of berries: ['hearts', 'stars', 'horseshoes', 'clovers']
+value of melons: ['hearts', 'stars', 'horseshoes', 'clovers']
+```
+
+Re-read through the code slowly, and take notes about how the modification to `berries` affects the value that `melons` points to.
+
+`berries` and `melons` both point to the same object. We modify the object with object ID `4369938880` with `berries.append("clovers")`. We see that the value of `melons` (which points to the object `4369938880`) is also modified.
+
+## Applied in Function Arguments
+
+When we pass in values into a function call, we should have awareness about if we are passing a mutable object or immutable object.
+
+As we've learned about parameters in functions, parameters are local variables.
+
+Now, with this lesson's vocabulary, we might say parameters initially _point_ to the same object of whatever argument is passed into a function call.
+
+### Mutable Objects Passed Into Functions
+
+When we pass a mutable object into a function, **that function can modify the object passed in**.
+
+Read through the code below. Take notes about how many variables there are, how many objects in memory there are, when modifications to those objects are made, and what the result is.
+
+```python
+def add_clovers(charms):
+    print("==== Appending 'clovers' to charms ====")
+    print("id of charms:", id(charms))
+    charms.append("clovers")
+
+berries = ["hearts", "stars", "horseshoes"]
+print("==== Before calling add_clovers ====")
+print("id of berries:", id(berries))
+print("value of berries:", berries)
+
+add_clovers(berries)
+
+print("==== After calling add_clovers ====")
+print("id of berries:", id(berries))
+print("value of berries:", berries)
+```
+
+When we run the above code, we may see output like this:
+
+```
+==== Before calling add_clovers ====
+id of berries: 4460399680
+value of berries: ['hearts', 'stars', 'horseshoes']
+==== Appending 'clovers' to charms ====
+id of charms: 4460399680
+==== After calling add_clovers ====
+id of berries: 4460399680
+value of berries: ['hearts', 'stars', 'horseshoes', 'clovers']
+```
+
+We should observe:
+
+1. `berries` points to a mutable object
+1. When we call `add_clovers`
+   - We pass in `berries`
+   - `charms` points to the same mutable object that `berries` does
+1. The object that `charms` points to is modified with `charms.append()`
+1. Because `charms` was modified, so was `berries`
+
+### Immutable Objects Passed Into Functions
+
+When we pass an immutable object into a function, that function cannot modify the object passed in. Any "modifications" are likely reassignment.
```

Comment:
Suggest emphasis on immutable since it's easy to lose track of which we're talking about.
Set up the following example.

```suggestion
When we pass an _immutable_ object into a function, that function cannot modify the object passed in. Any "modifications" are likely reassignment.

Recall that reassignment replaces a variable's reference with a reference to the new object. Since parameters are local variables, only the local reference is replaced. The external variable is left unchanged, as a brief example will show.
```